src/lexer.c

```diff
@@ -0,0 +1,111 @@
+#include "lexer.h"
+
+#include <err.h>
+#include <ctype.h>
+#include <string.h>
+
+static void quoting_reset(struct quoting_state *q)
+{
+	memset(q, 0, sizeof(*q));
+}
+
+static bool inside_quoting(const struct quoting_state *q)
+{
+	return q->backslashed == true;
+}
+
+int lexer_init(struct lexer *l, FILE *input)
+{
+	if (charstream_init(&l->stream, input) < 0) {
+		warnx("%s: failed to init charstream", __FUNCTION__);
+		return -1;
+	}
+
+	quoting_reset(&l->quoting);
+
+	l->cur = NULL;
+
+	return 0;
+}
+
+static void discard_line(struct lexer *l)
+{
+	char c;
+
+	do {
+		c = charstream_read(&l->stream);
+	} while (c != CHARSTREAM_EOF && c != '\n');
+}
+
+static int lexer_consume_char(struct lexer *l, struct token *tok)
+{
+	char c = charstream_read(&l->stream);
+	bool is_quoted = inside_quoting(&l->quoting);
+
+	/* step_quoting */
+	if (l->quoting.backslashed > 0)
+		l->quoting.backslashed--;
+
+	if (c == CHARSTREAM_EOF) {
+		token_delimit(tok);
+		return 0;
+	}
+
+	if (!is_quoted && c == '\\') {
+		l->quoting.backslashed = 1;
+		return 0;
+	}
+
+	if (!is_quoted && c == '#') {
+		discard_line(l);
+		return 0;
+	}
+
+	if (!is_quoted && isblank(c)) {
+		token_delimit(tok);
+		return 0;
+	}
+	
+	if (!is_quoted && c == '\n') {
+		token_delimit(tok);
+		return 0;
+	}
+	
+	if (token_append(tok, c) < 0)
+		return -1;
+
+	return 0;
+}
+
+/**
+ * Read chars from the charstream owned by @l until a token is delimited.
+ * Think of this as an iterator over the input tokens.
+ *
+ * @return token read by the lexer (the token is owned by the lexer)
+ */
```

Comment:
Shouldn't we put the doc in the header?

Reply:
As you want.

I put it in the source file because when looking at code you usually look at source files. Also ctags brings you to the source file, not the header so I thought it was more useful to see docs when jumping to a function definition.

Reply:
I think having the doc in headers is better. You have to scroll less between functions to find the documentation, and it's not hindered by all the static functions that the user cannot access anyway